### ISSUE TITLE
Search for (and translate) across all translation domains

### ISSUE THREAD

**Issue by madsoliver** [PrestaShop/PrestaShop#24385]
**The case**
The new translation system is great in many ways, and from a translators point of view, as well as for the developers, it makes sense to separate translation into `domains`.

However, for a merchant starting a PrestaShop with a bought theme and a bunch of bought modules, one might be lucky that the most is translated, except a few phrases that (s)he wants to translate. But in order to translate s(he) is met with a translation system that separates the translation into _domains_ that for her/him is the same thing: _the shop_.  It's rather intimidating and confusing, especially if s(he) tries to search in one place (`domain`) but cannot find the phrase, and then needs to search the second and third place (`domain`) in order to find and change that wording. I don't want to be the one that tells the merchant to [go learn about translation domains](https://devdocs.prestashop.com/1.7/development/internationalization/translation/translation-domains/), cause that ain't going to happen for my clients for sure 😅

Also, sometimes even third party module or theme developers does mistakes and places the wordings into the wrong domains, then searching is even harder.

**Possible solution**
Site-wide / domain-wide search. 
Wouldn't it be great if, when you just have some very few phrases you want to translate, then you could do a search across all translation domains? This search would be slower because it's alot more to search amongst, but, surely much faster than navigating between the different translation domains to try the search many places. It's a good tradeoff. My guess is that the site-wide / domain-wide search is the one that is going to be used most by merchants. For translators and developers _maybe_ we would go the more specific route and chose the correct domain and search there.
Site-wide / domain-wide search should be an extra option, not a replacement.

**Downside**
It's gonna be slow, or it's gotta be cached.

#### Additional context
Discussed in [_PrestaShop Project Public Demo #4_:](https://www.youtube.com/watch?v=VKb39T_6WHU&ab_channel=PrestaShopDevelopers)
_Would be slow_ according to @matks and @sowbiba 
_Possible if we cache it_ according to @eternoendless 
_Makes sense from users point of view_ according to @eternoendless

**Comment by eternoendless** [comment]
To search the original wordings, we'd need to load the full default catalogue: Core + Themes + Modules. Should be super slow the first time but could be made faster if put in cache.

Searching in translated wordings should be less costly because the sources are known: core xlf + module xlf + module php (legacy modules) + theme xlf + database.

The tricky part is editing, in particular with Core wordings used by themes, because you'd need to have one instance of the wording for the core and one instance for every theme using it.

(Note: this makes me think that perhaps it would be easier for everyone if we merged FO + BO + mails + others in a single page, then one page for each module, one for each theme. Mail template modification should be another separate page, apart from translations.)

**Comment by hibatallahAouadni** [comment]
Hello @onklmaps 

Thank you for your suggestion. The Product Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

PS: Thanks @eternoendless for your quick reply :wink: 

Thank you

**Comment by madsoliver** [comment]
> The tricky part is editing, in particular with Core wordings used by themes, because you'd need to have one instance of the wording for the core and one instance for every theme using it.

I think just linking to the instances from the search results, and then translating there (like today) is sufficient. The whole point IMO is to find translations that sometimes are hard to find (because there is so many places to look). Compare it to `grep -r "where did I put that config???" /` in your filesystem. You wouldn't expect it to be fast, and you wouldn't use it every time, but it sure is helpful when you don't know where to start looking.

**Comment by eternoendless** [comment]
Of course, as you know, I fully agree with the value of this feature. It's the implementation I'm worrying about :)

**Comment by boo-code** [comment]
Some numbers for the "it's gonna be slow" part, since the thread has the objection but not its size.

The core default catalogue in 9.2 is **164 XLIFF files, 8,589 trans-units, 3.0 MB on disk** - and that is one locale. Modules add their own on top, and a theme adds one more set. So the corpus a cross-domain search would have to load is a few tens of thousands of units, not the millions the phrase "search everything" tends to suggest. That is well within reach of a single cached pass; @eternoendless' "super slow the first time, faster if cached" reads as accurate rather than prohibitive.

Worth noting how the current search is shaped, because it is not far off already. `TreeBuilder::makeTranslationArray()` walks the **whole** default catalogue of the selected provider and filters each unit with `dataContainsSearchWord()` against the default wording, the XLIFF value and the database value. So the search is already a full linear scan with an in-PHP filter; it is only the *provider* that is scoped to one group. A cross-domain search is the same loop over more providers, not a different mechanism.

Which leaves the part @eternoendless flagged as the real difficulty, and it is a data-model question rather than a performance one: the same wording can exist under several providers - a core string reused by a theme - and each needs its own editable instance, because the database catalogue is keyed by domain and theme. A flat result list has to carry that context through to the save, or it will write the edit to the wrong instance.

That suggests a smaller first step than the feature as described: a **read-only** cross-domain search that returns each hit with its provider and domain and links to the existing editing screen positioned on it. It answers the merchant's actual problem in this issue - "I cannot find where this phrase lives" - without touching the write path at all, and it makes the caching question the only remaining one. Editing in place can follow once the instance-per-provider model is settled.

No patch prepared: the read-only shape above is small, but it is still a new back-office screen, and which of the two shapes to build is the product decision this issue has been waiting on.
